src/main/java/org/opensearch/knn/indices/ModelDao.java

```diff
@@ -0,0 +1,263 @@
+        public void put(KNNEngine knnEngine, byte[] modelBlob, ActionListener<IndexResponse> listener) {
+            String base64Model = Base64.getEncoder().encodeToString(modelBlob);
+
+            Map<String, Object> parameters = ImmutableMap.of(
+                    KNNConstants.KNN_ENGINE, knnEngine.getName(),
+                    KNNConstants.MODEL_BLOB_PARAMETER, base64Model
+            );
+
+            IndexRequestBuilder indexRequestBuilder = client.prepareIndex(MODEL_INDEX_NAME, "_doc");
+            indexRequestBuilder.setSource(parameters);
+
+            put(indexRequestBuilder, listener);
+        }
+
+        private void put(IndexRequestBuilder indexRequestBuilder, ActionListener<IndexResponse> listener) {
+            if (!isCreated()) {
+                throw new IllegalStateException("Cannot put model in index before index has been initialized");
+            }
+
+            // Fail if the id already exists. Models are not updateable
+            indexRequestBuilder.setOpType(DocWriteRequest.OpType.CREATE);
+            indexRequestBuilder.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+            indexRequestBuilder.execute(listener);
+        }
+
+        @Override
+        public byte[] get(String modelId) throws ExecutionException, InterruptedException {
+            if (!isCreated()) {
+                throw new IllegalStateException("Cannot get model \"" + modelId + "\". Model index does not exist.");
+            }
+
+            /*
+                GET /<model_index>/<modelId>?source_includes=<model_blob>&_local
+            */
+            GetRequestBuilder getRequestBuilder = new GetRequestBuilder(client, GetAction.INSTANCE, MODEL_INDEX_NAME)
+                    .setId(modelId)
+                    .setFetchSource(KNNConstants.MODEL_BLOB_PARAMETER, null)
+                    .setPreference("_local");
+            GetResponse getResponse = getRequestBuilder.execute().get();
+
+            Object blob = getResponse.getSourceAsMap().get(KNNConstants.MODEL_BLOB_PARAMETER);
+
+            if (blob == null) {
+                throw new IllegalArgumentException("No model available in \"" + MODEL_INDEX_NAME + "\" index with id \""
+                        + modelId + "\".");
+            }
+
+            return Base64.getDecoder().decode((String) blob);
+        }
+
+        private String getMapping() throws IOException {
+            URL url = ModelDao.class.getClassLoader().getResource(MODEL_INDEX_MAPPING_PATH);
+            if (url == null) {
+                throw new IllegalStateException("Unable to retrieve mapping for \"" + MODEL_INDEX_NAME + "\"");
+            }
+
+            return Resources.toString(url, Charsets.UTF_8);
+        }
+
+        @Override
+        public void delete(String modelId, ActionListener<DeleteResponse> listener) {
+            if (!isCreated()) {
+                throw new IllegalStateException("Cannot delete model \"" + modelId + "\". Model index does not exist.");
```

Comment:
If the Model index is deleted which means model is deleted right? Why should this be an exception?

Reply:
My thinking for throwing an exception is that a user should not try to delete something that isnt there. I can switch to a log statement and return.

Btw, this function is subject to change. For instance, what happens when someone tries to delete a model that is in use by an index? This could cause problems. I will need to think it through a little more once I start implementing Model Index management APIs.